### PR TITLE
Replace k-means++ CPU bottleneck with a `random::discrete` prim

### PR DESCRIPTION
**Pull request by Nyrio**
Currently, k-means has a CPU bottleneck in its k-means++ initialization, especially when the dataset is tall and thin. At each step of the k-means++ initialization, the distances to the closest cluster centroid are copied to the host, and candidates centroids are selected using [`std::discrete_distribution`](https://en.cppreference.com/w/cpp/numeric/random/discrete_distribution). This distribution reduces and scans the weights, which is an expensive operation if the dataset has many rows, and should be done on GPU.

For the test of this new primitive, I use a small number of weights and a large number of samples and compare the actual vs expected histogram using a tolerance of 4*sigma where sigma is the standard deviation computed from the number of samples and the smallest non-zero weight. I don't know any good way to correctly test the primitive with a very large number of weights and a small number of samples, but I'm open to suggestions.

### CODE REVIEW

**Comment by Nyrio** [comment]
Sorry for the absolute mess in the diff and automatic labels / review requests, but `branch-23.02` is behind `branch-22.12` by 22 commits...

**Comment by Nyrio** [comment]
Reverting target branch for now to `branch-22.12` to make reviews easier.

**Comment by Nyrio** [comment]
Nsight Systems timelines for comparison.

before
2022-11-22_kmeansplusplus_before

---

after
2022-11-22_kmeansplusplus_after

File: cpp/include/raft/random/detail/rng_device.cuh [hunk 1]
```diff
@@ -681,6 +681,40 @@ __global__ void rngKernel(DeviceState<GenType> rng_state,
...
 }
 
+template <typename GenType, typename OutType, typename WeightType, typename IdxType>
+__global__ void sample_with_replacement_kernel(DeviceState<GenType> rng_state,
+                                               OutType* out,
+                                               const WeightType* weights_csum,
+                                               IdxType sampledLen,
+                                               IdxType len)
+{
+  // todo(lsugy): warp-collaborative binary search
```
Comment (cjnolet) [inline cpp/include/raft/random/detail/rng_device.cuh:691]: Can you create an issue for this and reference it here?

**Comment by cjnolet** [comment]
> I don't know any good way to correctly test the primitive with a very large number of weights and a small number of samples, but I'm open to suggestions.

Could we do some little tricks here like setting most of the weights to be a very small value and then setting only a few weights to be much larger values and then comparing the unique set of sampled indices against the expected histogram to see that the unique list of indices that were sampled are all peaks in the expected histogram?

**Review by cjnolet — CHANGES_REQUESTED** [review]
The changes look good, though aside from adding the test for the smaller sample sizes w/ larger weights, we should add a quick usage example for the docs.

File: cpp/include/raft/random/rng.cuh [hunk 2]
```diff
@@ -703,6 +703,33 @@ void laplace(const raft::handle_t& handle,
   detail::laplace(rng_state, ptr, len, mu, scale, handle.get_stream());
 }
 
+/**
+ * @brief Generate random integers, where the probability of i is weights[i]/sum(weights)
+ *
```
[the remark this answers is not in the document]
Comment (Nyrio) [reply cpp/include/raft/random/rng.cuh:708]: I've added one, please let me know if you're happy with it.

**Comment by Nyrio** [comment]
I have consolidated the tests with small `sampled_len` / large `len`.

**Review by cjnolet — APPROVED** [review]
LGTM. I'm pre-approved, but we will need to change the target branch to 23.02 before this is merged.

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**